add export option for pdf, svg, png to plugin API

rolandlo opened the pull request:
This PR (when finished) closes #3565

Export already works. In the plugin you just need to add a line like

```lua 
app.export({["outputFile"] = "/home/USERNAME/Test.pdf", ["range"] = "2-5; 7", ["background"] = "all", ["progressiveMode"] = false})
```

(for pdf) or 

``` lua
app.export({["outputFile"] = "/home/USERNAME/Test.svg", ["range"] = "2-5; 7", ["background"] = "all"})
```

(for svg) or

``` lua
app.export({["outputFile"] = "/home/USERNAME/Test.png", ["range"] = "2-5; 7", ["background"] = "all", ["pngDpi"] = 300})
```
(for png). It's still a hack, since I mostly copied code from `XournalMain.cpp` (`exportPdf` and `exportImg` functions), but probably already usable.
 
The filename generation will be left to the plugin. I have added a "xoppFilename" field to the `app.getDocumentStructure()` table, so that only some string manipulations are needed.

**Edit:** Updated the information and removed the "pngDpi" field in the svg case, since it does not apply there.

jfab-123 commented:
Sorry for the ignorance but I don't really understand what this means.Is this a plugin? Where do I "add"  that line of code? How do I load it in xournalpp? Where can I read this about all of this?

rolandlo commented:
> Sorry for the ignorance but I don't really understand what this means.Is this a plugin? Where do I "add" that line of code? How do I load it in xournalpp? Where can I read this about all of this?

Ok, I should have explained. I have only extended the plugin API. You will have to write a plugin to use the extended API. How to write a plugin is explained in https://xournalpp.github.io/guide/plugins/plugins/#plugin-structure
For testing purposes you can also add that line to an existing plugin. For example modify the file `/usr/share/xournalpp/plugins/Example/main.lua` by adding the line at the beginning of the `exampleCallback` function (line 18). After activating the example plugin in menu `Plugin > Plugin manager` and restarting Xournal++ you can invoke the plugin via the shortcut "CTRL+T"

jfab-123 commented:
I get the following error when pressing ```<c-t>``` : 
![image](https://user-images.githubusercontent.com/81712003/141658843-cdc6f7e7-d8f4-4276-87ff-47a54922cca3.png)

rolandlo commented:
Did you check out the pull pull request and build from source? Have you ever built Xournal++ from source? The building process is described on [this wiki page](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md)

jfab-123 commented:
I have not, but I will read about it, thanks!

rolandlo commented:
@jfab-123 What OS do you have? We can also create an installer for a PR for some OS'es (Windows, macOS, Debian, Ubuntu 18.04, Ubuntu 20.04)

jfab-123 commented:
> @jfab-123 What OS do you have? We can also create an installer for a PR for some OS'es (Windows, macOS, Debian, Ubuntu 18.04, Ubuntu 20.04)

I have arch linux

jfab-123 commented:
I'm back. I built from source and it works! (I think). Whenever I press ```<c-t>```, the file is generated. However it is always a blank pdf file and it always has the same name. I guess this was just a test for testing purposes. 
It would be great if the shortcut saved the file, exported it with the same name and in the directory the file is in.

rolandlo commented:
The `Export` plugin I have added with the third commit should now export automatically to the same directory, 
where `/path/to/file/name.xopp` becomes `/path/to/file/name_export.pdf` (or svg/png).

When creating a new xopp-file from the command line, it should currently be created with an absolute file path to work properly.

jfab-123 commented:
Awesome! It works! (almost)
I don't know how the export functionality was implemented, but the svg export seems corrupted every time it is generated. When generating a blank test.xopp and then exporting it with the plugin, I get: 
```
/home/jfab/Documents/pbd/figures/anothertest_export.svg:1: parser error : Start tag expected, '<' not found
�PNG
^
/home/jfab/Documents/pbd/figures/anothertest_export.svg:1: parser error : Start tag expected, '<' not found
�PNG
^
ink_file_open: '/home/jfab/Documents/pbd/figures/anothertest_export.svg' cannot be opened!
InkscapeApplication::document_open: Failed to open: /home/jfab/Documents/pbd/figures/anothertest_export.svg
ConcreteInkscapeApplication::on_open: failed to create document!
```
And in latex I get the following error:
```
test.tex|12 warning| Package svg Warning: The export with Inkscape failed for file `figures//anothertest_export.svg' Troubleshooting: Please check in the log file how the invocation of Inkscape took place and try to execute it yourself in the terminal
```
This issue is not present when exporting to pdf nor png

jfab-123 commented:
Also, now that the plugin is created, will it be merged into the official app? Or do I have to use the xournal++ that I built from source? Will xournalpp be updated with this new feature?

rolandlo commented:
> Awesome! It works! (almost) I don't know how the export functionality was implemented, but the svg export seems corrupted every time it is generated. When generating a blank test.xopp and then exporting it with the plugin, if I try to open the svg with inkscape I get:
> 
> ```
> /home/jfab/Documents/pbd/figures/anothertest_export.svg:1: parser error : Start tag expected, '<' not found
> �PNG
> ^
> /home/jfab/Documents/pbd/figures/anothertest_export.svg:1: parser error : Start tag expected, '<' not found
> �PNG
> ^
> ink_file_open: '/home/jfab/Documents/pbd/figures/anothertest_export.svg' cannot be opened!
> InkscapeApplication::document_open: Failed to open: /home/jfab/Documents/pbd/figures/anothertest_export.svg
> ConcreteInkscapeApplication::on_open: failed to create document!
> ```
> 
> And in latex I get the following error:
> 
> ```
> test.tex|12 warning| Package svg Warning: The export with Inkscape failed for file `figures//anothertest_export.svg' Troubleshooting: Please check in the log file how the invocation of Inkscape took place and try to execute it yourself in the terminal
> ```
> 
> This issue is not present when exporting to pdf nor png. Also, exporting the usual way doesn't generate any problems

I fixed it. Please try again.

rolandlo commented:
> Also, now that the plugin is created, will it be merged into the official app? Or do I have to use the xournal++ that I built from source? Will xournalpp be updated with this new feature?

For the moment you will have to use the version you built from source. I think there is a good chance that the PR will (after some improvements and proper review) be merged into the official app. It will take some time, though. When it is merged (maybe in a couple of weeks), you will have it in the nightly releases (unstable version). Only with release of Xournal++ 1.2 (which will take some months) it will get into the stable version.

jfab-123 commented:
Sounds good! Thanks for the help!

personalizedrefrigerator commented on src/plugin/luapi_application.h:
```diff
@@ -867,6 +876,132 @@ static int applib_getDisplayDpi(lua_State* L) {
+        // exportPdf("test.xopp", outputFile, range, bgType, progressiveMode);
+        GFile* file = g_file_new_for_commandline_arg(outputFile);
+
+        XojPdfExport* pdfe = XojPdfExportFactory::createExport(doc, nullptr);
```
We're deleting `XojPdfExport` soon after we're creating it. Let's consider using a `std::unique_ptr` here!

rolandlo replied:
Done

personalizedrefrigerator commented on src/plugin/luapi_application.h:
```diff
@@ -867,6 +876,132 @@ static int applib_getDisplayDpi(lua_State* L) {
+ * Exports the current document as a pdf or as a svg or png image
+
+ * Example 1:
+ * app.export({["outputFile"] = "Test.pdf", ["range"] = "2-5; 7", ["background"] = "none", ["progressiveMode"] = true})
```
A quick documentation comment that describes what `progressiveMode` is could be helpful here! (Alternatively, this could be documented on the website/in an example plugin).

rolandlo replied:
Done (in Example 1)

personalizedrefrigerator reviewed (approved):
From a quick glance through the code (aside from some _very minor_ comments), this looks good to me!

rolandlo commented:
I have refactored the exporting functions in order to avoid code duplication. Moreover I have applied your review comments @personalizedrefrigerator 

Now it's time for testing. @jfab-123 If you have time, please test again after pulling the latest changes.

rolandlo commented:
@bhennion 
I have added auto-export of changed pages to the plugin as suggested by you and discussed in [this issue](https://github.com/xournalpp/xournalpp/pull/2618#issuecomment-1002321149). It is not polished by any means, but already works decently (if you don't change document or delete pages). To try it out

0) Checkout this PR
1) Make sure you have the Lua `lgi`-module installed
2) Start with a new document and save it in a (preferably) empty folder
3) Open that folder in your file manager (for observation)
4) Start auto-export from menu `Plugin` 
5) Write/draw on the page, insert new pages and write/draw on any of them. Observe the  png's created and updated

rolandlo commented:
I have decided to split up the PR. This is the first part, that introduces the export functionality to the plugin API. The second part will include the auto-exporting. Some non-trivial decisions will have to be made for the second part, while this first part seems ready to be merged.

rolandlo commented:
Merging in 48 hours if no objections are made.
